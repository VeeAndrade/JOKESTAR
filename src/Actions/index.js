export const updateUser = (username) => ({
  type: 'UPDATE_USER',
  username
});

export const setJoke = (joke) => ({
  type: 'SET_JOKE',
  joke
});

export const favoriteJoke = (joke) => ({
  type: 'FAVORITE_JOKE',
  joke
});

export const addFetchParams = (stateObj) => ({
  type: 'ADD_PARAMS',
  stateObj
});