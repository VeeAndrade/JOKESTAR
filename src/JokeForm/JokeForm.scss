.joke-form-wrapper {
  height: 90%;
  display: flex;
  align-items: center;
  align-self: center;
  justify-content: center;
  width: 95%;
}

.joke-form {
  display: flex;
  flex-direction: column;
  height: 90%;
  width: 90%;
  align-items: center;
  justify-content: center;
}

.joke-type-section,
.joke-flags-section,
.form-category-section,
.submit-form-button {
  // border: solid 1px pink;
  height: 20%;
  width: 80%;
  color: white;
  display: flex;
  justify-content: center;
  align-items: center;
}

.form-label {
  font-size: 1.4em;
  margin-right: 10px;
}

.category-select-elem {
  font-size: 1em;
  height: 2rem;
  width: 57%;

  &:hover {
    cursor: pointer;
  }
}

.submit-form-button {
  // border: solid 2px orange;
}

.submit-specifications {
  height: 35%;
  width: 75%;
  border-radius: 3px;
  font-size: 1.1em;
  border: none;

  &:hover {
    cursor: pointer;
    background-color: rgb(0, 255, 136);
  }
}

.input-checkboxes:hover {
  cursor: pointer;
}

.input-checkboxes {
  // height: 2em;
  // width: 2em;
}

.checkbox-span {
  font-size: 1.1em;
  margin-right: 10px;
}