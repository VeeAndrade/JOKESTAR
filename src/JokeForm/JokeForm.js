import React, { Component } from 'react';
import { getSpecificJoke } from '../apiCalls';
import { connect } from 'react-redux';
import './JokeForm.scss';
import { setJoke } from '../Actions'
import { addFetchParams } from '../Actions'


export class JokeForm extends Component {
  constructor() {
    super();
    this.state = {
      category: 'Any',
      blacklistFlags: [],
      type: []
    }
  }

  handleChange = (e) => {
    this.setState({[e.target.name]: [...this.state[e.target.name], e.target.value]})
  }

  handleCategory = (e) => {
    this.setState({category: e.target.value})
  }

  updateClickedButton = async (e) => {
    e.preventDefault()
    this.props.findClickedBtn(e)
    this.props.addFetchParams(this.state)
    let promiseJoke = await getSpecificJoke(this.state)
    this.props.updateJoke(promiseJoke)
  }

  render() {
    return (
      <section className='joke-form-wrapper'>
        <form className='joke-form'>
          <section className='form-category-section'>
          <label className='form-label'>Category:</label>
          <select className='category-select-elem' onChange={(event) => this.handleCategory(event)} value={this.state.category}>
            <option className='category-options' value='Any'>Any</option>
            <option className='category-options' value='Dark'>Dark</option>
            <option className='category-options' value='Miscellaneous'>Miscellaneous</option>
          </select>
          </section>
          <section className='joke-type-section'>
          <label className='form-label'>Type:</label>
            <input type='checkbox' onClick={(event) => this.handleChange(event)} className='input-checkboxes' name='type' value='single'/><span className='checkbox-span'>Single</span>
            <input type='checkbox' onClick={(event) => this.handleChange(event)} className='input-checkboxes' name='type' value='twopart'/><span className='checkbox-span'>Two-part</span>
          </section>
          <section className='joke-flags-section'>
          <label className='form-label'>Blacklist Flags(optional):</label>
            <input type='checkbox' onClick={(event) => this.handleChange(event)} className='input-checkboxes' name='blacklistFlags' value='nsfw'/><span className='checkbox-span'>NSFW</span>
            <input type='checkbox' onClick={(event) => this.handleChange(event)} className='input-checkboxes' name='blacklistFlags' value='religious'/><span className='checkbox-span'>Religious</span>
            <input type='checkbox' onClick={(event) => this.handleChange(event)} className='input-checkboxes' name='blacklistFlags' value='political'/><span className='checkbox-span'>Political</span>
            <input type='checkbox' onClick={(event) => this.handleChange(event)} className='input-checkboxes' name='blacklistFlags' value='racist'/><span className='checkbox-span'>Racist</span>
            <input type='checkbox' onClick={(event) => this.handleChange(event)} className='input-checkboxes' name='blacklistFlags' value='sexist'/><span className='checkbox-span'>Sexist</span>
          </section>
          <section className='submit-form-button'>
            <button className='submit-specifications' onClick={(event) => this.updateClickedButton(event)} value='getJoke'>Find Me a Joke!</button>
          </section>
        </form>
      </section>
    )
  }
}

export const mapStateToProps = state => ({
  joke: state.joke,
  fetchParams: state.fetchParams
})

export const mapDispatchToProps = dispatch => ({
  updateJoke: joke => dispatch(setJoke(joke)),
  addFetchParams: options => dispatch(addFetchParams(options))
})

export default connect(mapStateToProps, mapDispatchToProps)(JokeForm);