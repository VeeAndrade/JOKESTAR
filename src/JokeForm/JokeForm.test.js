import React from 'react';
import { JokeForm, mapStateToProps, mapDispatchToProps } from './JokeForm';
import { shallow } from 'enzyme';
import { setJoke, addFetchParams } from '../Actions'

describe('JokeForm', () => {

  let wrapper

  beforeEach(() => {
    wrapper = shallow( <JokeForm
    joke={{id: 42, setup: 'how much does a polarbear weigh?', delivery: 'enough to bresk the ice'}}
    fetchParams={{
      category: 'Any',
      blacklistFlags: [],
      type: []
    }}
    />)
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should unpdate state when handle change is invoked', () => {
    const mockState = {
      category: 'Any',
      blacklistFlags: [],
      type: []
    }
    const mockEvent = {
        target:
        {
          name: 'type',
          value: 'single'
        }
      }

      expect(wrapper.state()).toEqual(mockState);
      wrapper.instance().handleChange(mockEvent)
      expect(wrapper.state('type')).toEqual(['single'])
  });

  it('should update the category when handleCategory is called', () => {
    const mockState = {
      category: 'Any',
      blacklistFlags: [],
      type: []
    }
    const mockEvent = {
      target: {
        name: 'category',
        value: 'Dark'
      }
    }
    expect(wrapper.state('category')).toEqual('Any')
    wrapper.instance().handleCategory(mockEvent)
    expect(wrapper.state('category')).toEqual('Dark')
  })
}); 

describe('mapStateToProps', () => {
  const mockState = {
    joke: {id: 42, setup: 'how much does a polarbear weigh?', delivery: 'enough to bresk the ice'},
    fetchParams: {category: 'Any',
    blacklistFlags: [],
    type: []}
  }
  const expected = {
     joke:{id: 42, setup: 'how much does a polarbear weigh?', delivery: 'enough to bresk the ice'},
    fetchParams:{
      category: 'Any',
      blacklistFlags: [],
      type: []
    }
  }
  const mappedProps = mapStateToProps(mockState)
  expect(mappedProps).toEqual(expected);
});

describe('mapDispatchToProps', () => {
  it('should call updateJoke when update clicked button is clicked', () => {
    const mockResponse = {
      id: 42,
      setup: 'how much does a polarbear weigh?',
      delivery: 'enough to bresk the ice'
    }
    const mockDispatch = jest.fn();
    const actionToDispatch = setJoke(mockResponse)
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.updateJoke(mockResponse);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
  });

  it('should addFetchParams when update clicked button is called', () => {
    const mockResponse = {
      category: 'Any',
      blacklistFlags: [],
      type: []
    }
    const mockDispatch = jest.fn();
    const actionToDispatch = addFetchParams(mockResponse)
    const mappedProps = mapDispatchToProps(mockDispatch)
    mappedProps.addFetchParams(mockResponse)
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
  });
});