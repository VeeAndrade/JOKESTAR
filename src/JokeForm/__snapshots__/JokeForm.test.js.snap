// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JokeForm should match the snapshot 1`] = `
<section
  className="joke-form-wrapper"
>
  <form
    className="joke-form"
  >
    <section
      className="form-category-section"
    >
      <label
        className="form-label"
      >
        Category:
      </label>
      <select
        className="category-select-elem"
        onChange={[Function]}
        value="Any"
      >
        <option
          className="category-options"
          value="Any"
        >
          Any
        </option>
        <option
          className="category-options"
          value="Dark"
        >
          Dark
        </option>
        <option
          className="category-options"
          value="Miscellaneous"
        >
          Miscellaneous
        </option>
      </select>
    </section>
    <section
      className="joke-type-section"
    >
      <label
        className="form-label"
      >
        Type:
      </label>
      <input
        className="input-checkboxes"
        name="type"
        onClick={[Function]}
        type="checkbox"
        value="single"
      />
      <span
        className="checkbox-span"
      >
        Single
      </span>
      <input
        className="input-checkboxes"
        name="type"
        onClick={[Function]}
        type="checkbox"
        value="twopart"
      />
      <span
        className="checkbox-span"
      >
        Two-part
      </span>
    </section>
    <section
      className="joke-flags-section"
    >
      <label
        className="form-label"
      >
        Blacklist Flags(optional):
      </label>
      <input
        className="input-checkboxes"
        name="blacklistFlags"
        onClick={[Function]}
        type="checkbox"
        value="nsfw"
      />
      <span
        className="checkbox-span"
      >
        NSFW
      </span>
      <input
        className="input-checkboxes"
        name="blacklistFlags"
        onClick={[Function]}
        type="checkbox"
        value="religious"
      />
      <span
        className="checkbox-span"
      >
        Religious
      </span>
      <input
        className="input-checkboxes"
        name="blacklistFlags"
        onClick={[Function]}
        type="checkbox"
        value="political"
      />
      <span
        className="checkbox-span"
      >
        Political
      </span>
      <input
        className="input-checkboxes"
        name="blacklistFlags"
        onClick={[Function]}
        type="checkbox"
        value="racist"
      />
      <span
        className="checkbox-span"
      >
        Racist
      </span>
      <input
        className="input-checkboxes"
        name="blacklistFlags"
        onClick={[Function]}
        type="checkbox"
        value="sexist"
      />
      <span
        className="checkbox-span"
      >
        Sexist
      </span>
    </section>
    <section
      className="submit-form-button"
    >
      <button
        className="submit-specifications"
        onClick={[Function]}
        value="getJoke"
      >
        Find Me a Joke!
      </button>
    </section>
  </form>
</section>
`;
